### PR TITLE
ENH: Added a `PercentFormatter` class to `matplotlib.ticker`

### PR DESCRIPTION
**Pull request by madphysicist**
This is suitable for formatting axes labels as percentages. It has some nice features like being able to convert from arbitrary data scales to percents, a customizable percent symbol and either automatic or manual control over the decimal points.

The original motivation came from my own work as well as these two Stack Overflow questions:

  - http://stackoverflow.com/questions/26294360/pyplot-using-percentage-on-x-axis
  - http://stackoverflow.com/questions/31357611/format-y-axis-as-percent

### CODE REVIEW

**Comment by madphysicist** [comment]
@mdboom While I realize that you have a script to do this, it is really disconcerting to see your name pop up fractions of a second after I create a PR :)

File: lib/matplotlib/ticker.py [hunk 1]
```diff
@@ -914,6 +1009,17 @@ class EngFormatter(Formatter):
...
 
     def __init__(self, unit="", places=None):
+        """
+        Initializes an engineering notation formatter.
+
+        `unit` is a string containing the abbreviated name of the unit,
+        suitable for use with single-letter representations of powers of
+        1000. For example, 'Hz' or 'm'.
+
+        `places` is the percision with which to display the number,
```
Comment (efiring) [inline lib/matplotlib/ticker.py]: precision

File: lib/matplotlib/ticker.py [hunk 2]
```diff
@@ -108,33 +108,38 @@
...
+    Default formatter for scalars: autopick the format string
 
 :class:`LogFormatter`
-    formatter for log axes
+    Formatter for log axes
 
+:class:`EngFormatter`
+    Format labels in engineering notation
+
+:class:`PerentFormatter`
```
Comment (efiring) [inline lib/matplotlib/ticker.py]: Percent

File: lib/matplotlib/ticker.py [hunk 3]
```diff
@@ -973,6 +1078,87 @@ def format_eng(self, num):
...
+    `mx` parameter. `mx` is the data value that corresponds to 100%.
+    Percentages are computed as ``x / mx * 100``. So if the data is
+    already scaled to be percentages, `mx` will be 100. Another common
+    situation is where `max` is 1.0.
+    """
+    def __init__(self, mx=100, decimals=None, symbol='%'):
+        """
+        Initializes the formatter.
+
+        `max` is the data value that corresponds to 100%. `symbol` is
```
Comment (efiring) [inline lib/matplotlib/ticker.py]: This needs to match the name in the signature (i.e., 'max' needs to be the kwarg).

**Comment by madphysicist** [comment]
I am not familiar enough with Python 2.7 to immediately understand why the following is not working:

    s = '{x:0.{decimals}f}'.format(x=x, decimals=decimals)

Are nested formats not allowed?

File: lib/matplotlib/ticker.py [hunk 4]
```diff
@@ -973,6 +1078,87 @@ def format_eng(self, num):
...
+    """
+    Format numbers as a percentage.
+
+    How the number is converted into a percentage is determined by the
+    `mx` parameter. `mx` is the data value that corresponds to 100%.
+    Percentages are computed as ``x / mx * 100``. So if the data is
+    already scaled to be percentages, `mx` will be 100. Another common
+    situation is where `max` is 1.0.
+    """
+    def __init__(self, mx=100, decimals=None, symbol='%'):
```
Comment (efiring) [inline lib/matplotlib/ticker.py]: I think people will rarely want to specify `max`, so I would put it after the `decimals` kwarg.

File: lib/matplotlib/ticker.py [hunk 5]
```diff
@@ -973,6 +1078,87 @@ def format_eng(self, num):
...
+        else:
+            decimals = self.decimals
+        s = '{x:0.{decimals}f}'.format(x=x, decimals=decimals)
+
+        if self.symbol:
+            return s + self.symbol
+        return s
+
+    def convert_to_pct(self, x):
+        return x / self.max * 100.0
```
Comment (efiring) [inline lib/matplotlib/ticker.py]: I like to use parentheses to make expressions like this clearer; or multiply by 100 first.

**Comment by efiring** [comment]
There are so many cleanup changes here that I recommend breaking this into two PRs: one for the cleanups, and a second one for the new feature.

File: lib/matplotlib/ticker.py (hunk 4 above)
Comment (madphysicist) [inline lib/matplotlib/ticker.py]: I disagree. The most common parameter to modify would be `max`. This is intended for data going from 0 to 1, or as in the case of the SO question, from 0 to 5. I would expect the `decimals` argument to generally remain `None` assuming I have done a decent job with the automated version.

File: lib/matplotlib/ticker.py (hunk 5 above)
Comment (madphysicist) [inline lib/matplotlib/ticker.py]: Doing both. I agree that it looks awkward.

**Comment by madphysicist** [comment]
No problem with that. Is there an easy way to do that without closing this PR?

File: lib/matplotlib/ticker.py [hunk 6]
```diff
@@ -1016,7 +1016,7 @@ def __init__(self, unit="", places=None):
         suitable for use with single-letter representations of powers of
         1000. For example, 'Hz' or 'm'.
 
-        `places` is the percision with which to display the number,
+        `places` is the prrcision with which to display the number,
```
Comment (efiring) [inline lib/matplotlib/ticker.py]: Oops!

**Comment by QuLogic** [comment]
> @mdboom While I realize that you have a script to do this, it is really disconcerting to see your name pop up fractions of a second after I create a PR :)

That's https://waffle.io/

> No problem with that. Is there an easy way to do that without closing this PR?

Create a new branch starting on this one, rebase it to include only cleanup, and open a new PR for it. When that's merged, rebase this one on top of master and drop all the cleanup from here.

**Comment by efiring** [comment]
I think you would make a separate branch on your repo for the cleanups, and make a PR from that.  Then, on your pctFormatter branch, isolate the PercentFormatter part with additional commits, and use 'git rebase -i' to squash it down to a single commit.  Force-push that to your github repo, and it will replace the present commits in the present PR.

**Comment by madphysicist** [comment]
I did the split (#6253) and fixed the Py2.7 error. Turns out Py3 automatically converts acceptable format precisions into an int. Py2 does not.

The only potentially unrelated change I kept here was moving `__all__` to the top.

**Comment by madphysicist** [comment]
Metabolized the docs of `EngFormatter.format_eng` from #6253 into this commit to get rid of sphinx errors.

**Comment by madphysicist** [comment]
Error on setup, not in build. I'm pretty sure everything is passing now.

**Comment by madphysicist** [comment]
I think that this commit is complete (ready for further review). Same goes for #6253.

The failure in Appveyor is the sporadic image comparison failure. It is not related to the code I added as far as I can tell.

**Comment by madphysicist** [comment]
Squawk.

**Comment by mdboom** [comment]
Cc: @efiring.

**Comment by efiring** [comment]
In reading the new code, I find that I waste a lot of time trying to figure out what `d` really is.  I am comfortable with short variable names like `x` when their meaning is obvious from the context; but `d` is not obvious here--especially since some places it is in unscaled data units, and other places in percent.  So maybe use `dspan` or `xspan`, and `pcspan`, or something like that?  And maybe `xmax` instead of `mx`?  Readability matters, both in the tests and in the active code.

**Comment by madphysicist** [comment]
I changed `max` to `xmax` throughout, and changed `d` to `display_range` and `scaled_range` as appropriate. I think that does make a big improvement to readability.

On a side-note, I got `d` from `OldScalarFormatter`, originally thinking to reuse its `pprint` method. I don't think that `d` needs to be changed at this point, but it is a possibility. `d` also appears in at least one of the `Locator`s.

File: lib/matplotlib/tests/test_ticker.py [hunk 7]
```diff
@@ -366,6 +366,39 @@ def test_formatstrformatter():
...
+        (17.0, None, '%', 1, 8.5, '6%'),
+        (17.0, None, '%', 1, 8.4, '5.9%'),
+        (5, None, '%', -100, 0.000001, '-2000.00000%'),
+        # Check percent symbol
+        (1.0, 2, None, 1.2, 100, '120.00'),
+        (75, 3, '', 50, 100, '66.667'),
+        (42, None, '^^Foobar$$', 21, 12, '50.0^^Foobar$$'),
+    )
+    for xmax, decimals, symbol, x, display_range, expected in test_cases:
+        yield _percent_format_helper, xmax, decimals, symbol, x, display_range, expected
```
Comment (efiring) [inline lib/matplotlib/tests/test_ticker.py]: pep-8 failure here: line too long.  You could use parentheses; or you could use
```python
    for case in test_cases:
        yield (_percent_format_helper,) + case
```
I haven't tested, but I think that would work and would be more readable.

**Comment by efiring** [comment]
In addition to the pep-8 failure noted above there is a real failure that I have not tracked down (probably you will spot it instantly).  Once those are taken care of, I think this will be functionally ready.  It's a nice contribution. You could neaten it up by rebasing to squash the commits.

**Comment by madphysicist** [comment]
I went with

    for case in test_cases:
        yield (_percent_format_helper, *case)

and also fixed the other stupid typo. Will push with squash soon.

**Comment by madphysicist** [comment]
And now I know why I should have listened the first time.

**Comment by madphysicist** [comment]
I don't think the Travis failure is related to my code.

**Comment by efiring** [comment]
Thank you, @madphysicist!

**Comment by madphysicist** [comment]
No problem. Thanks for looking over and accepting my PR!

**Pull request merged**